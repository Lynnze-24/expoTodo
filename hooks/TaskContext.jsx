import React, { createContext, useContext, useState,useEffect} from 'react'
import uuid from 'react-native-uuid';
import AsyncStorage from '@react-native-async-storage/async-storage';

const TaskContext = createContext();


export function useTaskContext(){
    return useContext(TaskContext)
}

export function TaskProvider({children,mock,val}) {
    // mock and val are for testing
    const initial = mock?mock : [];
    const [task, setTask] = useState(initial);
    const fetchTasks = mock||val? ()=>{return}:async() => {
        let localTasks= await AsyncStorage.getItem('TaskAHL') || null;
        if(localTasks){
            setTask(JSON.parse(localTasks));
        }
    }// second function is for testing 

    
    useEffect(()=> {
        fetchTasks()
    
    }, [])
    
    const addToTask = async(item)=>{
        let newTask = [...task,item]
        setTask(newTask)
        AsyncStorage.setItem('TaskAHL',JSON.stringify(newTask))
    }
    const removeFromTask = async(id)=>{
        let newTask = task.filter(x=> x.id!== id);
        setTask(newTask)
        AsyncStorage.setItem('TaskAHL',JSON.stringify(newTask))
    }

    const toggleCheck = async(id) =>{
        let newTask = [...task];
       let changedItem = newTask.find(x=> x.id=== id);
       changedItem.isChecked = !changedItem.isChecked;
       setTask(newTask);
       AsyncStorage.setItem('TaskAHL',JSON.stringify(newTask))
    }
    
    const value = val || {task,addToTask,removeFromTask,toggleCheck};
    
    return (
        <TaskContext.Provider value={value}>
        {children}
      </TaskContext.Provider>
    )
}