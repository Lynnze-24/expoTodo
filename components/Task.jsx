import React,{useState} from 'react'
import { View, Text,StyleSheet,Alert,TouchableOpacity } from 'react-native'
import { Checkbox, IconButton } from 'react-native-paper';
import cusTheme from '../theme/Theme';
import { MaterialCommunityIcons } from '@expo/vector-icons'; 
import { useTaskContext } from '../hooks/TaskContext';
import { useDevice } from '../hooks/Device';







const Task = ({name,isChecked,id,mock}) => {
//    mock is for testing cases
   
    const {toggleCheck,removeFromTask} = useTaskContext();
    const {isWeb} = useDevice()

    const confirmDelete = () =>{
        Alert.alert('', 'Are you sure to delete this item?', [
            {
              text: 'Cancel',
              onPress: () => console.log('Cancel Pressed'),
              style: 'cancel',
            },
            { text: 'OK', onPress: () => removeFromTask(id) },
          ]);
    }

    const deleteWeb = ()=> removeFromTask(id)



    
   
    return (
        <View testID='task' style={styles.container}>
            <Checkbox.Android testID='checkbox' onPress={()=> toggleCheck(id) }  status={isChecked?'checked':'unchecked'}  color={cusTheme.primary}/>
            <Text testID='todoText' style={[styles.text,isChecked && styles.completed]}>{name}</Text>
            {mock?(<TouchableOpacity testID='delete'  onPress={isWeb?deleteWeb: confirmDelete}></TouchableOpacity>)
            :(<MaterialCommunityIcons  onPress={isWeb?deleteWeb: confirmDelete} style={styles.icon} name="delete" size={24} color={cusTheme.primary} />)}
        </View>
    )
}

const styles = StyleSheet.create({
    container:{
        flexDirection:'row',
        alignItems:'center',
        paddingVertical:3,
        borderColor:cusTheme.primary,
        borderRadius:10,
        borderWidth:1,
        marginVertical:5,
    },
    text:{
        fontSize:18,
    },
    completed:{
        textDecorationLine:'line-through'
    },
    icon:{
        position:'absolute',
        right:10,
        top:8,
    },
})

export default Task;
