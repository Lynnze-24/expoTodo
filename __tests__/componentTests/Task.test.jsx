import { render, fireEvent } from '@testing-library/react-native';
import { TaskProvider } from '../../hooks/TaskContext';

import Task from '../../components/Task';
import renderer from 'react-test-renderer';
import {describe} from '@jest/globals';



// mock  functions 
    const addToTask = jest.fn();
    const task = jest.fn();
    const removeFromTask = jest.fn();
    const toggleCheck = jest.fn();

it('renders correctly', () => {
        const comp = renderer.create((<TaskProvider >
             <Task mock={true} name={'go shopping'} isChecked={true} id={'aaa'} key={'aaa'}/>
        </TaskProvider>)).toJSON();
        expect(comp).toMatchSnapshot();
});

it('should render all child components',()=>{
    
        const { getByTestId}  = render(
               <TaskProvider val={{task,addToTask,removeFromTask,toggleCheck}}>
                    <Task mock={true} name={'go shopping'} isChecked={true} id={'aaa'} key={'aaa'}/>
               </TaskProvider>
                );
        getByTestId('checkbox')
        getByTestId('todoText')
        getByTestId('delete')
})

it('should render childs correctly according to props',()=>{
   
    
    const {queryByTestId}  = render(
           <TaskProvider val={{task,addToTask,removeFromTask,toggleCheck}}>
               <Task mock={true} name={'go shopping'} isChecked={true} id={'aaa'} key={'aaa'}/>
           </TaskProvider>
            );

        // checkbox
            
      expect(queryByTestId('checkbox')).toHaveProp('accessibilityState',{
          checked:true,
          disabled:false
      });  // ischecked is true so accessibilityState.checked should be true
      expect(queryByTestId('todoText')).toHaveTextContent('go shopping') 
      expect(queryByTestId('todoText')).toHaveStyle({
                                             textDecorationLine:'line-through'
                                        }) // should have completed text style since it is checked
});

it('should toggle check when click checkbox',()=>{
    
    const { getByTestId}  = render(
           <TaskProvider val={{task,addToTask,removeFromTask,toggleCheck}}>
                <Task mock={true} name={'go shopping'} isChecked={true} id={'aaa'} key={'aaa'}/>
           </TaskProvider>
            );
    const checkbox = getByTestId('checkbox')
    fireEvent.press(checkbox);
    expect(toggleCheck).toHaveBeenCalledWith('aaa')
   
})





describe('platform tests', () => {

    let Platform;
    beforeEach(() => {
        Platform = require('react-native').Platform;
    });

    describe('ios tests', () => {
        beforeEach(() => {
            Platform.OS = 'ios';
        });

        it('should use correct function according to Platform',()=>{
    
            const { getByTestId}  = render(
                   <TaskProvider val={{task,addToTask,removeFromTask,toggleCheck}}>
                        <Task mock={true} name={'go shopping'} isChecked={true} id={'aaa'} key={'aaa'}/>
                   </TaskProvider>
                    );
                    fireEvent.press( getByTestId('delete'))
                             
                    expect(removeFromTask).not.toHaveBeenCalledWith('aaa');
        })
    });

    describe('android tests', () => {
        beforeEach(() => {
            Platform.OS = 'android';
        });

        it('should use correct function according to Platform',()=>{
    
            const { getByTestId}  = render(
                   <TaskProvider val={{task,addToTask,removeFromTask,toggleCheck}}>
                        <Task mock={true} name={'go shopping'} isChecked={true} id={'aaa'} key={'aaa'}/>
                   </TaskProvider>
                    );
                    
          
                    fireEvent.press( getByTestId('delete'))
                             
                    expect(removeFromTask).not.toHaveBeenCalledWith('aaa');
        })
    });

    describe('web tests', () => {
        beforeEach(() => {
            Platform.OS = 'web';
        });

        it('should use correct function according to Platform',()=>{
    
            const { getByTestId}  = render(
                   <TaskProvider val={{task,addToTask,removeFromTask,toggleCheck}}>
                        <Task mock={true} name={'go shopping'} isChecked={true} id={'aaa'} key={'aaa'}/>
                   </TaskProvider>
                    );
                   
          
                    fireEvent.press(getByTestId('delete'))
                             
                    expect(removeFromTask).toHaveBeenCalledWith('aaa');
        })
    });

});