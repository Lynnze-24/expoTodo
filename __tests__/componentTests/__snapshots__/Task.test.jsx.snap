// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "borderColor": "#120311",
      "borderRadius": 10,
      "borderWidth": 1,
      "flexDirection": "row",
      "marginVertical": 5,
      "paddingVertical": 3,
    }
  }
  testID="task"
>
  <View
    accessibilityLiveRegion="polite"
    accessibilityRole="checkbox"
    accessibilityState={
      Object {
        "checked": true,
        "disabled": false,
      }
    }
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "overflow": "hidden",
        },
        Object {
          "borderRadius": 18,
          "height": 36,
          "padding": 6,
          "width": 36,
        },
      ]
    }
    testID="checkbox"
  >
    <View
      collapsable={false}
      style={
        Object {
          "transform": Array [
            Object {
              "scale": 1,
            },
          ],
        }
      }
    >
      <Text />
      <View
        style={
          Array [
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
            Object {
              "alignItems": "center",
              "justifyContent": "center",
            },
          ]
        }
      >
        <View
          collapsable={false}
          style={
            Object {
              "borderColor": "#120311",
              "borderWidth": 0,
              "height": 14,
              "width": 14,
            }
          }
        />
      </View>
    </View>
  </View>
  <Text
    style={
      Array [
        Object {
          "fontSize": 18,
        },
        Object {
          "textDecorationLine": "line-through",
        },
      ]
    }
    testID="todoText"
  >
    go shopping
  </Text>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID="delete"
  />
</View>
`;
